[WIP] feat: Üublishing of DataSetMetaData via mqtt json

wlkrm opened the pull request:
TODO

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open62541/open62541?pullRequest=5118) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/open62541/open62541?pullRequest=5118) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open62541/open62541?pullRequest=5118) it.</sub>

andreasebner commented:
Thank you for the contribution @wlkrm! Currently the appveyor build is failing: 5>Compiling...
6>..\src\pubsub\ua_pubsub_networkmessage_json.c(246) : error C[236](https://ci.appveyor.com/project/open62541/open62541/builds/43558754/job/ltrkkjjhc0m7xeri#L236)1: initialization of 'count' is skipped by 'default' label
6>        ..\src\pubsub\ua_pubsub_networkmessage_json.c(217) : see declaration of 'count'
6>..\src\pubsub\ua_pubsub_networkmessage_json.c(246) : error C2361: initialization of 's' is skipped by 'default' label
6>        ..\src\pubsub\ua_pubsub_networkmessage_json.c(171) : see declaration of 's

It would be great if you could rebase your changes and search for a solution. I’m looking forward to merge your changes, but currently the coverage will decrease significantly because your changes do not contain test cases. 
@NoelGraf could you please review this PR?

wlkrm commented:
I will add the missing test cases.

NoelGraf commented on include/open62541/server_pubsub.h:
```diff
@@ -287,6 +288,7 @@ typedef struct {
 typedef struct {
     UA_String name;
     UA_PublishedDataSetType publishedDataSetType;
+    UA_Boolean sendViaWriterGroupTopic;
```
Is the insertion of the field necessary?
And what's the field for?

wlkrm replied:
The field allows for selective batching. DataSetMessages of DataSetWriters can be batched by their common WriterGroup using the `UA_BrokerDataSetWriterTransportDataType` of the WriterGroup according to Part 14. PublishedDataSets of Writers grouped by a common WriterGroup with this flag set will be sent out batched under the Data-Topic-Queue defined in the transportSettings (`UA_BrokerDataSetWriterTransportDataType.queueName`) of the WriterGroup. The reason for the implementation is, that my use-case gets simplified, if I receive DataSetMessages of multiple DataSetWriter grouped together through batching. Example JSONs and Topics of this can be found here (https://github.com/umati/Showcase/blob/pubsub/PubSub.md#topic-structure).

wlkrm commented on src/server/ua_server.c:
```diff
@@ -822,7 +823,7 @@ UA_Server_run_iterate(UA_Server *server, UA_Boolean waitInternal) {
         UA_PubSubConnection *ps = connection;
         if(ps && ps->channel->yield){
-            ps->channel->yield(ps->channel, 0);
+            ps->channel->yield(ps->channel, 10);
```
Yes, that was the reason. I omitted this change in the rebase.

wlkrm commented:
@NoelGraf, thank you for your comments. I rebased the PR and tests for the encoding. I was not really sure how to properly test the MQTT-Part, as the tests would have to spin up a broker? Please review again.

jpfr commented on CMakeLists.txt:
```diff
@@ -1276,7 +1291,7 @@ else()
     if(UA_ENABLE_PUBSUB)
         list(APPEND UA_FILE_DATATYPES ${PROJECT_SOURCE_DIR}/tools/schema/datatypes_pubsub.txt)
-        if(UA_ENABLE_PUBSUB_INFORMATIONMODEL)
+        if(UA_ENABLE_PUBSUB_INFORMATIONMODEL OR UA_ENABLE_PUBSUB_MQTT_METADATA)
```
Why does metadata publishing require the nodeset xml?
Add a short comment to explain.

wlkrm replied:
In fact it is not necessary. I removed it.

jpfr reviewed (commented):
Looks really good overall.
Please make this configurable with a configuration flag, not a build flag.

MQTT+JSON is not used in environments where every kilobyte counts.
So we can afford to have this compiled in and keep the complexity of the build options low.

wlkrm commented:
@jpfr I removed the compile time switches and replaced them by a per DataSetWriter config field. A DataSetWriter can now be configured (https://github.com/wlkrm/open62541/blob/34156d549bab3e00d9d71b715e99dec81ffa6f81/include/open62541/server_pubsub.h#L599) to publish the DataSetMetaData using the corresponding transport layer (e.g. JSON-MQTT, which is the only metadata transport implemented in this PR).

jpfr commented:
> I removed the compile time switches and replaced them by a per DataSetWriter config field. A DataSetWriter can now be configured (https://github.com/wlkrm/open62541/blob/34156d549bab3e00d9d71b715e99dec81ffa6f81/include/open62541/server_pubsub.h#L599

If this is the place where the feature is enabled, add a comment that the feature requires JSON+MQTT.
Otherwise people will not understand what it does without looking at the code.

We can also log a warning if the feature is enabled without JSON+MQTT.

If there is a canonical section in the PubSub spec, also refer to that in the comment.

andreasebner commented on examples/pubsub/tutorial_pubsub_mqtt_publish.c:
```diff
@@ -272,7 +272,7 @@ addWriterGroup(UA_Server *server, char *topic, int interval) {
     transportSettings.content.decoded.data = &brokerTransportSettings;
 
-    writerGroupConfig.transportSettings = transportSettings;
+    UA_ExtensionObject_copy(&transportSettings, &writerGroupConfig.transportSettings);
```
Why is this coppy needed? The config structures should be deep-copied e.g. witihin "UA_Server_addWriterGroup"

wlkrm replied:
It is not needed. Removed it

andreasebner commented on examples/pubsub/tutorial_pubsub_mqtt_publish.c:
```diff
@@ -351,7 +353,7 @@ addDataSetWriter(UA_Server *server, char *topic) {
     transportSettings.content.decoded.data = &brokerTransportSettings;
 
-    dataSetWriterConfig.transportSettings = transportSettings;
+    UA_ExtensionObject_copy(&transportSettings, &dataSetWriterConfig.transportSettings);
```
Same as before.

wlkrm replied:
Done

andreasebner commented on examples/pubsub/tutorial_pubsub_mqtt_publish.c:
```diff
@@ -459,10 +461,12 @@ int main(int argc, char **argv) {
-
+    
+    UA_NodeId fieldNodeId;
+    memset(&fieldNodeId, 0, sizeof(UA_NodeId));
```
Please use UA_NodeId_init in the example.

wlkrm replied:
Done

andreasebner commented on examples/pubsub/tutorial_pubsub_mqtt_publish.c:
```diff
@@ -471,7 +475,20 @@ int main(int argc, char **argv) {
     addDataSetWriter(server, topic);
 
-    UA_Server_run(server, &running);
+    UA_Server_run_startup(server);
+    size_t iteration = 0;
+    while(running) {
+        iteration++;
+        UA_Server_run_iterate(server, true);
+        if(iteration % 100 == 0) {
+            /* Trigger publishing of DataSetMetaData. In a real application we would have made 
+             * changes to the properties of the corresponding node at this point, 
+             * which results in an update of the MinorVersion of the DataSetMetaData's ConfigurationVersion
+             * and thus, a new DataSetMetaData message is sent. */ 
+            UA_Server_updateDataSetField(server, fieldNodeId, UA_TRUE, UA_FALSE);
+        }
+    }
+    UA_Server_run_shutdown(server);
```
This code-path is very complex and execution dependet. Please use a cyclic callback to e.g. publish the metada every 5 seconds.

wlkrm replied:
Done

andreasebner commented on include/open62541/server_pubsub.h:
```diff
@@ -314,6 +315,7 @@ typedef struct {
 typedef struct {
     UA_String name;
     UA_PublishedDataSetType publishedDataSetType;
+    UA_Boolean sendViaWriterGroupTopic;
```
Maybe a more precise name of this flag would help to understand it. In addition add a short inline comment.

wlkrm replied:
Renamed it to batchMessagesViaWriterGroupTopic and added the coment.

andreasebner commented on include/open62541/server_pubsub.h:
```diff
@@ -409,6 +411,9 @@ UA_Server_addDataSetField(UA_Server *server,
                           UA_NodeId *fieldIdentifier);
 
+void UA_EXPORT
+UA_Server_updateDataSetField(UA_Server *server, const UA_NodeId dsf, UA_Boolean minor, UA_Boolean major);
```
This function seems for me a little bit missunderstanding. Our API-Supports other "updateXX" functions, but these have a different purpose. Please think about a better name - maybe UA_Server_sendPubSubMetaData? In addition i can't get why you neeed the minor and major flags. The user should not decide if the version-number changes. It should be depended on PDS changes without user interaction.

wlkrm replied:
I renamed it to UA_Server_triggerDataSetFieldUpdate (https://github.com/wlkrm/open62541/blob/d5ce05ca1bda027fadb23789f277e7fa465900c1/include/open62541/server_pubsub.h#L418). The functions enables the user to inform the server that a DataSetField has changed. This allows to make multiple changes to the DataSetField, to avoid inconsistent states, e.g. if the user wants to make multiple changes "at once".  UA_Server_sendPubSubMetaData is not a suitable name imo, as the function does not really send MetaData.

andreasebner commented on src/pubsub/ua_pubsub_datasetmetadata_json.c:
```diff
@@ -0,0 +1,80 @@
+static const char* UA_DECODEKEY_DATASETWRITERNAME = "DataSetWriterName";
+
+UA_StatusCode
+UA_DataSetMetaData_encodeJson_internal(const UA_DataSetMetaData *src, void *c) {
```
Internal functions mostly file local / static.

wlkrm replied:
Done

andreasebner commented on src/pubsub/ua_pubsub_datasetmetadata_json.c:
```diff
@@ -0,0 +1,80 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ *
+ * Copyright (c) 2022 ISW (for umati and VDW e.V.) (Author: Moritz Walker)
+ */
```
The function isn't that long and i think it will not grow that much in the future. I would preffer to add it to the networkmessage layer.

wlkrm replied:
Done

andreasebner commented on src/pubsub/ua_pubsub_writer.c:
```diff
@@ -134,6 +135,7 @@ UA_StatusCode
 UA_PublishedDataSetConfig_copy(const UA_PublishedDataSetConfig *src,
                                UA_PublishedDataSetConfig *dst) {
     UA_StatusCode res = UA_STATUSCODE_GOOD;
+    dst->sendViaWriterGroupTopic = src->sendViaWriterGroupTopic;
```
Non nested fields are copyed by the memcpy afterwards.

wlkrm replied:
I removed this.

andreasebner commented on src/pubsub/ua_pubsub_writer.c:
```diff
@@ -473,6 +487,39 @@ addDataSetField(UA_Server *server, const UA_NodeId publishedDataSet,
     return result;
 }
 
+void
+UA_Server_updateDataSetField(UA_Server *server, const UA_NodeId dsf, UA_Boolean minor, UA_Boolean major) {
+    UA_DataSetField *currentField = UA_DataSetField_findDSFbyId(server, dsf);
+    if(!currentField) {
+        return;
+    }
+
+    if(currentField->configurationFrozen) {
+        UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                       "Update DataSetField failed. DataSetField is frozen.");
+        return;
+    }
+
+    UA_PublishedDataSet *pds =
+        UA_PublishedDataSet_findPDSbyId(server, currentField->publishedDataSet);
+    if(!pds) {
+        return;
+    }
+
+    if(pds->configurationFrozen) {
+        UA_LOG_WARNING(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                       "Update DataSetField failed. PublishedDataSet is frozen.");
+        return;
+    }
+
+    /* Update minor version of PublishedDataSet */
+    UA_UInt32 timeDifference = UA_PubSubConfigurationVersionTimeDifference();
+    if(minor)
+        pds->dataSetMetaData.configurationVersion.minorVersion = timeDifference;
+    if(major)
+        pds->dataSetMetaData.configurationVersion.majorVersion = timeDifference;
+}
+
```
This code does not match the logic implied by the function name. In addition the user can change the minor/major version directly. I would preffer to have only implicit changes.

wlkrm replied:
I removed the flags.

andreasebner commented on src/pubsub/ua_pubsub_writer.c:
```diff
@@ -890,6 +938,60 @@ addDataSetWriter(UA_Server *server,
+                        break;
+                }
+#ifdef UA_ENABLE_JSON_ENCODING
+                res = sendNetworkMessageMetadataJson(connection, &dataSetMetaData, &newDataSetWriter->config.dataSetWriterId, 1, &newDataSetWriter->config.transportSettings);
```
This means we send out the Metadata during the WG creation. Not sure if this is allowed if the state is not operational.

wlkrm replied:
MetaData are now published for the first time, when the transport is ready.
https://github.com/open62541/open62541/pull/5118/commits/63ab001e387136a30925936656b9e60aa3c113fb

wlkrm commented:
> > I removed the compile time switches and replaced them by a per DataSetWriter config field. A DataSetWriter can now be configured (https://github.com/wlkrm/open62541/blob/34156d549bab3e00d9d71b715e99dec81ffa6f81/include/open62541/server_pubsub.h#L599
> 
> If this is the place where the feature is enabled, add a comment that the feature requires JSON+MQTT. Otherwise people will not understand what it does without looking at the code.
> 
> We can also log a warning if the feature is enabled without JSON+MQTT.
> 
> If there is a canonical section in the PubSub spec, also refer to that in the comment.

I added the comment. And a warning: https://github.com/wlkrm/open62541/blob/63ab001e387136a30925936656b9e60aa3c113fb/src/pubsub/ua_pubsub_writer.c#L891
